input/witnessgen.go

```diff
@@ -7,87 +7,123 @@ import (
 	"github.com/btcsuite/btcd/wire"
 )
 
-// WitnessType determines how an output's witness will be generated. The
-// default commitmentTimeLock type will generate a witness that will allow
-// spending of a time-locked transaction enforced by CheckSequenceVerify.
-type WitnessType uint16
+// WitnessGenerator represents a function that is able to generate the final
+// witness for a particular public key script. Additionally, if required, this
+// function will also return the sigScript for spending nested P2SH witness
+// outputs. This function acts as an abstraction layer, hiding the details of
+// the underlying script.
+type WitnessGenerator func(tx *wire.MsgTx, hc *txscript.TxSigHashes,
+	inputIndex int) (*Script, error)
+
+// WitnessType determines how an output's witness will be generated. This
+// interface can be implemented to be used for custom sweep scripts if the
+// pre-defined StandardWitnessType list doesn't provide a suitable one.
+type WitnessType interface {
+	// String returns a human readable version of the WitnessType.
+	String() string
+	
+	// GenFunc will return a WitnessGenerator function that an output uses
+	// to generate the witness and optionally the sigScript for a sweep
+	// transaction.
+	GenFunc(signer Signer, descriptor *SignDescriptor) WitnessGenerator
```

Comment:
I feel `GenFunc` is not a very descriptive name. Could we call it just `(Get)WitnessGenerator`?

Reply:
Renamed it to `WitnessGenerator` and removed the `Get` prefixes of the other methods.